### PR TITLE
feat: support authorization code flow

### PR DESCRIPTION
**Pull request by lucacavallaro**
### List of changes
1. Add `op-app` workspace, that expose an express application with `oidc-provider` configured to enable the [Authentication Code Flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowSteps)
2. Rewrite the `oidc-provider` adapter for persistence to use only Cosmos [Point Reads](https://devblogs.microsoft.com/cosmosdb/point-reads-versus-queries/) instead of cross-partition-query and non-efficient lookups.
3. Remove legacy code, and redundant `fp-ts` abstractions over `oidc-provider`
4. Revert back to TypeScript 5.4.3 to tempfix an issue with vscode and yarn https://github.com/yarnpkg/berry/issues/6270
5. Removed legacy `openid-provider` workspace

### What's missing
1. There are no `consent` and `login` screens, they will be implemented in subsequents changes (we could not use the old consent screen because it is web-based. Our now flow has a native consent screen).
2. HSM

### Motivation and context
Our `oidc-provider` implementation was tied to [Implicit Flow](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowSteps) that is no longer recommended.

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [x] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [x] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [ ] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### CODE REVIEW

**Comment by gunzip** [comment]
lgtm

File: apps/op-app/src/adapters/outbound/cosmos/oidc/adapter.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  async destroy(id: string): Promise<void> {
+    try {
+      await this.#container.item(id, id).delete();
+    } catch {
+      // these model have a ttl, so "destroy(id)" can fail silenty
+      // in case of cosmos error
+    }
+  }
+
+  async findByUid(uid: string): Promise<oidc.AdapterPayload | void> {}
```
Comment (gunzip) [inline apps/op-app/src/adapters/outbound/cosmos/oidc/adapter.ts:68]: why empyt methods? I'd a comment

**Review by gunzip — COMMENTED** [review]
good job!

File: apps/op-app/src/adapters/outbound/cosmos/oidc/grantable-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      const { resource } = await this.#grantedsByGrantId
+        .item(grantId, grantId)
+        .read<{ payload: { id: string } }>();
+      if (resource) {
+        const id = resource.payload.id;
+        await this.destroy(id);
+      }
+    } catch {
+      // these model have a ttl, so "destroy(id)" can fail silenty
+      // in case of cosmos error
```
Comment (Garma00) [inline apps/op-app/src/adapters/outbound/cosmos/oidc/grantable-adapter.ts:51]: Do we want to throw an error here?

File: apps/op-app/src/adapters/outbound/cosmos/oidc/session-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+export default class SessionAdapter extends Adapter {
+  #sessionsByUid: Container;
+
+  constructor(container: Container) {
+    super(container);
+    this.#sessionsByUid = container.database.container("sessions-by-uid");
+  }
+
+  async upsert(id: string, payload: oidc.AdapterPayload, ttl: number) {
```
Comment (Garma00) [inline apps/op-app/src/adapters/outbound/cosmos/oidc/session-adapter.ts:16]: I would add the return type as you did for other methods

**Review by Garma00 — COMMENTED** [review]
Awesome, can you add a `README.md` with at least some instruction to run this app locally?

File: apps/op-app/src/adapters/outbound/cosmos/oidc/grantable-adapter.ts (hunk 2 above)
Comment (lucacavallaro) [reply apps/op-app/src/adapters/outbound/cosmos/oidc/grantable-adapter.ts:51]: nope, as written in the comment we can fail silently here (throwing would stop the oidc flow)

File: apps/op-app/src/adapters/outbound/cosmos/oidc/adapter.ts (hunk 1 above)
Comment (lucacavallaro) [reply apps/op-app/src/adapters/outbound/cosmos/oidc/adapter.ts:73]: Done https://github.com/pagopa/io-fims/pull/53/commits/92cecdc7b5a631b57cc8c7eb4661c0d77272dbe4 ✅

**Comment by lucacavallaro** [comment]
> Awesome, can you add a `README.md` with at least some instruction to run this app locally?

Done in https://github.com/pagopa/io-fims/pull/53/files#diff-2649b33f45b9c6fb0fb8a9003c39129a9d1a113f1353a0c8cb9f9303abf280ad ✅

**Review by gunzip — COMMENTED** [review]
lgtm

**Review by Garma00 — APPROVED** [review]

**Pull request merged**